content/kubermatic/main/architecture/supported-providers/kubevirt/kubevirt.en.md

```diff
@@ -554,6 +555,73 @@ Follow the below steps to import the dashboard in Grafana:
+### KubeVirt upgrade
+
+tbd.
+
+### Cluster labels
+
+With KKP 2.22 KubeVirt CCM is upgraded from `v0.2.0` to `v0.4.0`.
+From `v0.3.0` on it requires [new labels](https://github.com/kubevirt/cloud-provider-kubevirt#prerequisites) on VMs to correctly route the traffic to services.
+
+Newly created VMs will have these labels automatically but for existing VMs and VMIs there is manual action required on your KubeVirt cluster.
+
+Add following labels
+```yaml
+cluster.x-k8s.io/cluster-name: <cluster-id>
+cluster.x-k8s.io/role: worker
+```
+to:
+
+* `VirtualMachine.metadata.labels`
+* `VirtualMachine.spec.template.metadata.labels`
+* `VirtualMachineInstance.metadata.labels`
+
+Additionally if there are already existing LoadBalancer Services in your cluster namespace on the KubeVirt cluster, you also need to manually add these labels to these Service's `spec.selector`
+
+#### Step by step example
+
+Patch VirtualMachine, VirtualMachineInstance, LoadBalander Service:
+```
+$ export KUBECONFIG=<kubeconfig of your KubeVirt cluster>
+$ export CLUSTERID=<your kkp user cluster id>
+$ cat << EOF > patch-vm.yaml
+metadata:
+  labels:
+    cluster.x-k8s.io/cluster-name: $CLUSTERID
+    cluster.x-k8s.io/role: worker
+spec:
+  template:
+    metadata:
+      labels:
+        cluster.x-k8s.io/cluster-name: $CLUSTERID
+        cluster.x-k8s.io/role: worker
+EOF
+$ kubectl patch -n cluster-$CLUSTERID vm <vm> --patch-file patch-vm.yaml --type=merge
+$ cat << EOF > patch-vmi.yaml
+metadata:
+  labels:
+    cluster.x-k8s.io/cluster-name: $CLUSTERID
+    cluster.x-k8s.io/role: worker
+EOF
+$ kubectl patch -n cluster-$CLUSTERID vmi <vmi> --patch-file patch-vmi.yaml --type=merge
+$ cat << EOF > patch-lb-svc.yaml
+spec:
+  selector:
+    cluster.x-k8s.io/cluster-name: $CLUSTERID
+    cluster.x-k8s.io/role: worker
+EOF
+$ kubectl patch -n cluster-$CLUSTERID svc <lb-svc> --patch-file patch-lb-svc.yaml --type=merge
+```
+
+### MachineDeployments
+
+MachineDeployments created with KKP 2.22 will have the field `clusterName` set with the cluster ID in the KubeVirt `cloudProviderSpec`,
+which is used by the machine-controller to set the cluster labels on VMs and VMIs. Any pre-existing MachineDeployments will not have this
+field set automatically to avoid rolling over all Machines. So in the scenario of scaling up pre-existing MachineDeployments for
+LoadBalancers to properly function, there are 2 possibilities:
+
+1. If it's no problem that all VMs get re-created, just set `MachineDeployment.spec.template.spec.providerSpec.value.cloudProviderSpec.clusterName` to your cluster ID.
+2. Otherwise you need to manually set the cluster labels described above on all VMs and VMIs that are being created while scaling the MachineDeployment up.
```

Comment:
Do we want to tell the customer to also update the MD (will be useful in case of md scaling for example) ?
And maybe warn him that this will imply that a new VM will be spawned ?
What do you think ?